File: 10-stacks-queues1/4-expression-evaluation/main.py
# Expression Evaluation - Python Example

def evaluate_postfix(expr):
    stack = []
    for token in expr.split():
        if token.isdigit():
            stack.append(int(token))
        else:
            b = stack.pop()
            a = stack.pop()
            if token == '+':
                stack.append(a + b)
            elif token == '-':
                stack.append(a - b)
            elif token == '*':
                stack.append(a * b)
            elif token == '/':
                stack.append(a // b)
    return stack[0]

expr = "3 4 2 * 1 5 - 2 3 ^ ^ / +"
print(f"Result of postfix expression: {evaluate_postfix(expr)}")

File: 10-stacks-queues1/4-expression-evaluation/main.java
// Expression Evaluation - Java Example

import java.util.Stack;

public class ExpressionEvaluation {

    public static int evaluatePostfix(String expr) {
        Stack<Integer> stack = new Stack<>();
        String[] tokens = expr.split("\\s+");

        for (String token : tokens) {
            if (token.matches("\\d+")) {
                stack.push(Integer.parseInt(token));
            } else {
                int b = stack.pop();
                int a = stack.pop();
                switch (token) {
                    case "+":
                        stack.push(a + b);
                        break;
                    case "-":
                        stack.push(a - b);
                        break;
                    case "*":
                        stack.push(a * b);
                        break;
                    case "/":
                        stack.push(a / b);
                        break;
                }
            }
        }
        return stack.pop();
    }

    public static void main(String[] args) {
        String expr = "3 4 2 * 1 5 - 2 3 ^ ^ / +";
        System.out.println("Result of postfix expression: " + evaluatePostfix(expr));
    }
}
